Adds support for Kong Workspaces

tjrivera opened the pull request:
Hi @hbagdi,

Starting small here but wanted to keep the ball rolling on getting more EE support in go-kong per #17.

This adds basic Workspace CRUD capabilities as well as a helper function to manage testing CE vs. EE.  

This does not include the management of Workspace Entities -- I'd like to keep PRs small if possible. 

Thanks,
Tyler

Commit:
feat: Adds workspace support

hbagdi commented on kong/workspace_service_test.go:
```diff
@@ -0,0 +1,84 @@
+	workspace, err = client.Workspaces.Update(defaultCtx, workspace)
+	assert.Nil(err)
+	assert.NotNil(workspace)
+	assert.Equal("teamA", *workspace.Name)
```
Let's add an assert her to confirm `.Comment` is actually populated.

Can you also please add something in the `.Config` and `.Meta` to confirm they are correctly populated.

hbagdi commented on kong/workspace_service.go:
```diff
@@ -0,0 +1,130 @@
+}
+
+// List fetches a list of all Workspaces in Kong.
+func (s *WorkspaceService) List(ctx context.Context) ([]*Workspace, error) {
```
Please match the signature of `List` method in other services. This should include ListOpt argument.

hbagdi commented on kong/workspace_service.go:
```diff
@@ -0,0 +1,130 @@
+	}
+
+	return workspaces, nil
+}
```
Please also add in `ListAll` method as well.

hbagdi commented:
Please rebase on latest master.

Commit:
allows testing across CE and EE

Commit:
spelling typo

Commit:
Adds support for workspace entities

hbagdi commented:
Tyler, can we please handle one endpoint at a time? Can you open a separate PR for workspace entity? That code requires more review it seems. The workspace code was almost ready to go but the entity work is blocking this PR.

hbagdi commented on kong/kong_test.go:
```diff
@@ -113,3 +121,22 @@ func runWhenKong(t *testing.T, semverRange string) {
+
+// runWhenEnterprise skips a test if the version
+// of Kong running is not enterprise edition.
+func runWhenEnterprise(t *testing.T) {
```
This function should take a semver range as well. There are new endpoints introduced in newer versions of enterprise product that won't be supported by old versions and that will lead to test breakage.

tjrivera replied:
I've added support for semver range. thanks

hbagdi commented on kong/kong_test.go:
```diff
@@ -98,6 +98,14 @@ func runWhenKong(t *testing.T, semverRange string) {
+		// A bit of a hack. Kong sometimes reports a version that
+		// does not match semver conventions. i.e. "1.5.0.1" will
+		// not parse properly.
+		v = strings.Join(strings.Split(v, ".")[:3], ".")
```
I'd recommend to implement this logic into a new method that can be re-used if need be. The previous logic was simple enough but with inclusion of this and the "rc" hack above that, we should probably refactor it. If you feel that's a scope creep to this PR and should rather be done separately, that's fine too.

tjrivera replied:
I think you're right, but I'm going to defer this refactor to a separate PR

hbagdi commented on kong/workspace_service_test.go:
```diff
@@ -0,0 +1,91 @@
+	err = client.Workspaces.Delete(defaultCtx, createdWorkspace.ID)
+	assert.Nil(err)
+}
+func TestWorkspaceServiceList(T *testing.T) {
```
You don't have test for List()

tjrivera replied:
added. thanks!

Commit:
removes workspace entity service

Commit:
adds workspace service list test

Commit:
updates runWhenEnterprise to accept a semver argument

It looks like workspaces were added in 0.33. I don't know if it's actually worthwhile to pass this into our workspace tests as I doubt anything below 0.33 is on the testing matrix.

Commit:
implements cleanVersionString in runWhenKong

hbagdi commented:
There was a leftover WorkspaceEntity in `kong/types.go`. I removed that and merged in manually.
Thank you!
Let's move to the next one.
